File: core/proxy/src/main/java/de/tomcory/heimdall/core/proxy/littleshoot/mitm/CertificateSniffingMitmManager.java
package de.tomcory.heimdall.core.proxy.littleshoot.mitm;

import java.nio.ByteBuffer;
import java.security.cert.Certificate;
import java.security.cert.X509Certificate;

import javax.net.ssl.SSLEngine;
import javax.net.ssl.SSLPeerUnverifiedException;
import javax.net.ssl.SSLSession;

import de.tomcory.heimdall.core.proxy.littleshoot.MitmManager;

import de.tomcory.heimdall.core.util.ByteUtils;
import io.netty.handler.codec.http.HttpRequest;
import timber.log.Timber;

/**
 * {@link MitmManager} that uses the common name and subject alternative names
 * from the upstream certificate to create a dynamic certificate with it.
 */
public class CertificateSniffingMitmManager implements MitmManager {

    private BouncyCastleSslEngineSource sslEngineSource;

    public CertificateSniffingMitmManager() throws RootCertificateException {
        this(new Authority());
    }

    public CertificateSniffingMitmManager(Authority authority)
            throws RootCertificateException {
        try {
            sslEngineSource = new BouncyCastleSslEngineSource(authority, true, true);
        } catch (final Exception e) {
            throw new RootCertificateException(
                    "Errors during assembling root CA.", e);
        }
    }

    public SSLEngine serverSslEngine(String peerHost, int peerPort) {
        return sslEngineSource.newSslEngine(peerHost, peerPort);
    }

    public SSLEngine serverSslEngine() {
        return sslEngineSource.newSslEngine();
    }

    public SSLEngine clientSslEngineFor(HttpRequest httpRequest, SSLSession serverSslSession) {
        try {
            X509Certificate upstreamCert = getCertificateFromSession(serverSslSession);
            // TODO store the upstream cert by commonName to review it later

            // A reasons to not use the common name and the alternative names
            // from upstream certificate from serverSslSession to create the
            // dynamic certificate:
            //
            // It's not necessary. The host name is accepted by the browser.
            //
            String commonName = getCommonName(upstreamCert);

            SubjectAlternativeNameHolder san = new SubjectAlternativeNameHolder();

            san.addAll(upstreamCert.getSubjectAlternativeNames());

            Timber.d("Subject Alternative Names: $san");
            return sslEngineSource.createCertForHost(commonName, san);

        } catch (Exception e) {
            throw new FakeCertificateException(
                    "Creation dynamic certificate failed", e);
        }
    }

    private X509Certificate getCertificateFromSession(SSLSession sslSession)
            throws SSLPeerUnverifiedException {
        Certificate[] peerCerts = sslSession.getPeerCertificates();
        Certificate peerCert = peerCerts[0];
        if (peerCert instanceof X509Certificate) {
            return (X509Certificate) peerCert;
        }
        throw new IllegalStateException(
                "Required java.security.cert.X509Certificate, found: "
                        + peerCert);
    }

    public String getCommonName(X509Certificate c) {
        Timber.d("Subject DN principal name: ${c.getSubjectDN().getName()}");
        Timber.d("DN " + c.getSubjectX500Principal().getName() + " hash is " + ByteUtils.bytesToHex(ByteBuffer.allocate(4).putInt(c.getSubjectX500Principal().hashCode()).array()));
        for (String each : c.getSubjectDN().getName().split(",\\s*")) {
            if (each.startsWith("CN=")) {
                String result = each.substring(3);
                Timber.d("Common Name: $result");
                return result;
            }
        }
        throw new IllegalStateException("Missed CN in Subject DN: "
                + c.getSubjectDN());
    }
}

File: core/vpn/src/main/java/de/tomcory/heimdall/core/vpn/mitm/CertificateSniffingMitmManager.kt
package de.tomcory.heimdall.core.vpn.mitm

import java.security.cert.X509Certificate
import javax.net.ssl.SSLEngine
import javax.net.ssl.SSLPeerUnverifiedException
import javax.net.ssl.SSLSession
import javax.net.ssl.SSLSocket

/**
 * MitmManager that uses the common name and subject alternative names
 * from the upstream certificate to create a dynamic certificate with it.
 */
class CertificateSniffingMitmManager(authority: Authority?) {

    private var sslEngineSource: SSLEngineSource? = try {
        if (authority != null) {
            SSLEngineSource(authority, trustAllServers = true, sendCerts = true)
        } else {
            null
        }
    } catch (e: Exception) {
        throw RootCertificateException("Error assembling SSLEngineSource with root CA", e)
    }


    fun createServerSSLEngine(peerHost: String?, peerPort: Int): SSLEngine? {
        return if(peerHost != null) sslEngineSource?.newSSLEngine(peerHost, peerPort) else createServerSSLEngine()
    }

    fun createServerSSLEngine(): SSLEngine? {
        return sslEngineSource?.newSSLEngine(true)
    }

    fun createServerSSLSocket(peerHost: String?, peerPort: Int): SSLSocket? {
        return if(peerHost != null) sslEngineSource?.newSSLSocket(peerHost, peerPort) else createServerSSLSocket()
    }

    fun createServerSSLSocket(): SSLSocket? {
        return sslEngineSource?.newSSLSocket()
    }

    /**
     * Call with serverSslEngine.getSession(). Creates an SSLEngine for the client side that
     * impersonates the remote server with a fake certificate.
     */
    fun createClientSSLEngineFor(serverSSLSession: SSLSession): SSLEngine {
        val engine = try {
            val upstreamCert = getCertificateFromSession(serverSSLSession)
            //TODO: store the upstream cert by commonName to review it later

            // A reasons to not use the common name and the alternative names
            // from upstream certificate from serverSslSession to create the
            // dynamic certificate:
            // It's not necessary. The host name is accepted by the browser.
            val commonName = getCommonName(upstreamCert)
            val san = SubjectAlternativeNameHolder()
            san.addAll(upstreamCert.subjectAlternativeNames)
            sslEngineSource!!.createCertForHost(commonName, san)
        } catch (e: Exception) {
            throw FakeCertificateException("Creation dynamic certificate failed", e)
        }

        engine.useClientMode = false

        return engine
    }

    @Throws(SSLPeerUnverifiedException::class)
    private fun getCertificateFromSession(sslSession: SSLSession): X509Certificate {
        val peerCerts = sslSession.peerCertificates
        val peerCert = peerCerts[0]
        if (peerCert is X509Certificate) {
            return peerCert
        }
        throw IllegalStateException("Required java.security.cert.X509Certificate, found: $peerCert")
    }

    private fun getCommonName(c: X509Certificate): String {
        val cnIndex = c.subjectDN.name.indexOf("CN=", ignoreCase = true)
        if (cnIndex == -1 || cnIndex + 3 >= c.subjectDN.name.length) {
            throw IllegalStateException("Missing CN in Subject DN: " + c.subjectDN)
        }
        val cnEndIndex = c.subjectDN.name.indexOf(",", cnIndex)
        return if (cnEndIndex == -1) {
            c.subjectDN.name.substring(cnIndex + 3).trim()
        } else {
            c.subjectDN.name.substring(cnIndex + 3, cnEndIndex).trim()
        }
    }

    //TODO: singleton isn't ideal here; it would be better to attach it to the VpnService lifecycle
    companion object {
        private var singleton: CertificateSniffingMitmManager? = null

        @JvmStatic
        fun createSingleton(authority: Authority): CertificateSniffingMitmManager? {
            singleton = CertificateSniffingMitmManager(authority)
            return singleton
        }

        @JvmStatic
        fun getSingleton(): CertificateSniffingMitmManager? {
            return singleton
        }
    }
}